File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
 vector<vector<vector<int>>> dp;

 int solve(int i, int buy, int k, vector<int>& prices) {
    if( i==prices.size() or k>=2)return 0;
    if (dp[i][buy][k] != -1)
            return dp[i][buy][k];

    if( buy){
         return dp[i][buy][k] = max( -prices[i]+ solve(i+1,0,k,prices),0+solve(i+1,1,k,prices));
    }
    else return dp[i][buy][k] = max(prices[i] + solve(i + 1, 1, k + 1, prices),
                solve(i + 1, 0, k, prices));
 }
    int maxProfit(vector<int>& prices) {
         int n = prices.size();
        dp = vector<vector<vector<int>>>(n,
              vector<vector<int>>(2, vector<int>(3, -1)));
              return solve(0, 1, 0, prices);
    }
};

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        memo = [[[-1] * 2 for _ in range(2)] for _ in range(n)]
        
        def solve(idx, buy, t):
            if idx == n or t == 2: return 0
            if memo[idx][buy][t] != -1: return memo[idx][buy][t]
            
            ans = 0
            if buy:
                buy_skip = solve(idx + 1, 1, t)
                buy_take = solve(idx + 1, 0, t) - prices[idx]
                
                ans = max(buy_skip, buy_take)
                
            else:
                sell_skip = solve(idx + 1, 0, t)
                sell_take = solve(idx + 1, 1, t + 1) + prices[idx]
                
                ans = max(sell_skip, sell_take)
                
            memo[idx][buy][t] = ans
            return ans
            
        return solve(0, 1, 0)